File: packages/backend/app/schemas/sharing.py
from datetime import datetime
from typing import Optional
from uuid import UUID
from pydantic import BaseModel


class ShareCreate(BaseModel):
    shared_with_github_login: Optional[str] = None  # None = public link
    permission: str = 'view'  # 'view' | 'comment'


class ShareResponse(BaseModel):
    id: UUID
    annotation_set_id: UUID
    shared_by: UUID
    shared_with: Optional[UUID] = None
    share_token: str
    permission: str
    created_at: datetime

    class Config:
        from_attributes = True


class AnnotationData(BaseModel):
    id: str
    set_id: str
    page_number: int
    type: str
    rects: list
    selected_text: Optional[str] = None
    note_content: Optional[str] = None
    color: Optional[str] = None


class AnnotationSetData(BaseModel):
    id: str
    pdf_id: str
    name: str
    color: str
    annotations: list[AnnotationData]


class SharedAnnotationsResponse(BaseModel):
    """Public response for /shared/annotations/{token} — no auth required."""
    shared_by_login: str
    shared_by_avatar: Optional[str]
    permission: str
    annotation_set: AnnotationSetData
    pdf_id: str
    pdf_title: str

File: packages/frontend/src/api/sharing.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { apiFetch } from './client';

export interface Share {
    id: string;
    annotation_set_id: string;
    shared_by: string;
    shared_with: string | null;
    shared_with_github_login?: string | null;
    share_token: string;
    permission: 'view' | 'comment';
    created_at: string;
}

interface ShareCreate {
    shared_with_github_login?: string;
    permission?: 'view' | 'comment';
}

interface AnnotationData {
    id: string;
    set_id: string;
    page_number: number;
    type: string;
    rects: Array<{ x: number; y: number; w: number; h: number }>;
    selected_text?: string | null;
    note_content?: string | null;
    color?: string | null;
}

interface SharedAnnotationsResponse {
    shared_by_login: string;
    shared_by_avatar: string | null;
    permission: string;
    annotation_set: {
        id: string;
        pdf_id: string;
        name: string;
        color: string;
        annotations: AnnotationData[];
    };
    pdf_id: string;
    pdf_title: string;
}

// Queries

export const useSharesForSet = (setId: string) =>
    useQuery<Share[]>({
        queryKey: ['shares', setId],
        queryFn: () => apiFetch(`/annotation-sets/${setId}/shares`),
        enabled: !!setId,
    });

/** Public — no auth required. Used on /shared/:token page. */
export const useSharedAnnotations = (token: string) =>
    useQuery<SharedAnnotationsResponse>({
        queryKey: ['shared-annotations', token],
        queryFn: () => apiFetch(`/shared/annotations/${token}`, { authRequired: false }),
        enabled: !!token,
        retry: false,
    });

// Mutations

export const useCreateShare = (setId: string) => {
    const qc = useQueryClient();
    return useMutation<Share, Error, ShareCreate>({
        mutationFn: (data) =>
            apiFetch(`/annotation-sets/${setId}/share`, {
                method: 'POST',
                body: JSON.stringify(data),
            }),
        onSuccess: () => qc.invalidateQueries({ queryKey: ['shares', setId] }),
    });
};

export const useRevokeShare = () => {
    const qc = useQueryClient();
    return useMutation<void, Error, string>({
        mutationFn: (shareId) => apiFetch(`/shares/${shareId}`, { method: 'DELETE' }),
        onSuccess: () => qc.invalidateQueries({ queryKey: ['shares'] }),
    });
};
